### ISSUE TITLE
[PSyData] Add support for PSyData nodes to Fortran visitor

### ISSUE THREAD

**Issue by arporter** [stfc/PSyclone#1168]
While working on #737 I've realised that there's no Visitor support for PSyData profiling nodes. This issue will add it.

**Comment by arporter** [comment]
On looking at this, I realise that a PSyData node is a high-level concept that will actually require quite a number of low-level/primitive PSyIR nodes to represent it. Therefore I will need to implement the `lower_to_language_level` method introduced by @sergisiso.

**Comment by arporter** [comment]
The PSyData API makes calls to type-bound procedures, e.g.

    call psyData1%postEnd()

Currently, the `Call.create()` method only accepts a `RoutineSymbol`. I think this means there are two problems: 1. should it also accept a `StructureReference/ArrayOfStructuresReference`? and 2. we probably need a new `RoutineMember` node to describe members of a structure that are routines. If @sergisiso agrees I shall open (another) new ticket and do that extension first.

**Comment by sergisiso** [comment]
Since this enters in the OO territory I avoided it for now and used CodeBlocks or hardcoded routine names with "%" as it makes me slightly worried if this will truly be language independent and/or opens the PSyIR to lots of complexities. But you may be right that we need something like what you propose.

Would a FortranCodeBlock work for the PSyData lower_to_language_level() for the time being? And if we had the proper PSyIR support would the PSyData supporting library be still Fortran-only? 

If you think we need it anyways I am happy to discuss what is the best way to implement this.

**Comment by arporter** [comment]
Thanks Sergi - I think a `CodeBlock` will work for now so I'll go with that approach.

**Comment by arporter** [comment]
Previously, the name of each profiled region was determined automatically (if not explicitly supplied) by counting the number of `ProfileNode`s in the tree. Perhaps we can add an annotation to the `CodeBlock` nodes being introduced to say that they were previously `ProfileNode`s?

**Comment by arporter** [comment]
Another day, another question: the profiling API adds a symbol of derived type to the generated code. This must have the `save` attribute because it stores state between calls of the invoke. In addition, since, internally, the PSyData wrapper keeps a pointer to this symbol, it must have the `target` attribute. @sergisiso am I right in thinking that we don't have either of those concepts in the PSyIR as yet?

**Comment by arporter** [comment]
I've realised that I can probably use `UnknownFortranType` for the symbols of derived type that need unsupported attributes.

**Comment by sergisiso** [comment]
You are right we don't have these concepts, and again we have to consider carefully if adding this concepts its worth the additional complexity. If for type-bound procedures I was a bit on the fence, I am less fond of these attributes.

If it is just to produce Fortran and no transformations or language conversion are needed here, then creating the UnknownFortranType (or letting a quick Fortran->fparser->frontend to create the UnknownFortranType declarations for you) is better for the time being.

**Comment by rupertford** [comment]
I would have thought that `save` represents a relatively generic 'lifetime' concept that would be needed in other contexts?. However, `target` is pretty Fortran-specific as far as I know so would hopefully be able to be derived from the fact that something is "pointing" to it, which means we need the concept of a pointer (which I assume we don't have?).

**Comment by sergisiso** [comment]
What I was thinking is that these attributes have an affect on:
- Device acceleration where there is separate memory and pointer/static may be invalid. (although the same is true for globals and this are managed well by the transformation that removes them)
- Dependency analysis (as @hiker mention each of this variations make the analysis more complex )

I have to admit that I was thinking more on terms of CodeBlocks, which we can treat them as something that the transformations will refuse or barriers that things could move in in-between CodeBlock but not across them. However this would be regular References just with an UnknownDataType, I don't know if we treat these safely (e.g. some transformation validate fails if they are found). 😟 

In the longer term it may be desirable to have as many concepts as possible as it allows transformations to manage them with more accurately depending on the case but it potentially adds more branching/logic/validate conditions in other places. I don't want to seem to be against adding them, if they make somebody use case easy that is a good reason to add them.

**Comment by arporter** [comment]
For the moment I'm pleased to say that our "support for things we don't support" (i.e. CodeBlocks and UnknownType) are working well in this case so there's no need to do extra work on the PSyIR yet...

**Comment by sergisiso** [comment]
#1177 added support for ProfileNode, but I am keeping this issue open until we verify if this is also solved for other PSyData nodes like ExtractNode.

Using the FortranWriter for GOcean would be a good test as it contains a few ExtractNode tests.
